Modules/posixmodule.c

```diff
@@ -5088,384 +5088,407 @@ os__path_splitroot_impl(PyObject *module, path_t *path)
-            PyErr_Clear();
-            Py_RETURN_FALSE;
-        }
-        return NULL;
+    DWORD fileDevType = GetFileType(hfile);
+    if ((fileDevType == FILE_TYPE_UNKNOWN && GetLastError()) ||
+        (fileDevType == FILE_TYPE_CHAR) ||
+        (diskOnly && fileDevType != FILE_TYPE_DISK))
+    {
+        return FALSE;
     }
 
-    Py_BEGIN_ALLOW_THREADS
-    if (_path.wide) {
-        if (_Py_GetFileInformationByName(_path.wide, FileStatBasicByNameInfo,
-                                         &statInfo, sizeof(statInfo))) {
-            if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)) {
-                slow_path = FALSE;
-                result = statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY;
-            } else if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
-                slow_path = FALSE;
-                result = 0;
-            }
-        } else if (_Py_GetFileInformationByName_ErrorIsTrustworthy(GetLastError())) {
-                    slow_path = FALSE;
-                    result = 0;
+    DWORD attributes, reparseTag;
+    if ((testedType != PY_IFREG && testedType != PY_IFDIR)) {
+        FILE_ATTRIBUTE_TAG_INFO info;
+        if (!GetFileInformationByHandleEx(hfile, FileAttributeTagInfo, &info,
+                                          sizeof(info)))
+        {
+            return FALSE;
         }
+        attributes = info.FileAttributes;
+        reparseTag = info.ReparseTag;
     }
-    if (slow_path) {
-        if (_path.fd != -1) {
-            hfile = _Py_get_osfhandle_noraise(_path.fd);
-            close_file = FALSE;
-        }
-        else {
-            hfile = CreateFileW(_path.wide, FILE_READ_ATTRIBUTES, 0, NULL,
-                                OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    else {
+        FILE_BASIC_INFO info;
+        if (!GetFileInformationByHandleEx(hfile, FileBasicInfo, &info,
+                                          sizeof(info)))
+        {
+            return FALSE;
         }
-        if (hfile != INVALID_HANDLE_VALUE) {
-            if (GetFileInformationByHandleEx(hfile, FileBasicInfo, &info,
-                                            sizeof(info)))
-            {
-                result = info.FileAttributes & FILE_ATTRIBUTE_DIRECTORY;
-            }
-            else {
-                result = 0;
-            }
-            if (close_file) {
-                CloseHandle(hfile);
-            }
+        attributes = info.FileAttributes;
+        reparseTag = 0;
+    }
+
+    if (attributes & FILE_ATTRIBUTE_REPARSE_POINT) {
+        switch (testedType) {
+        case PY_IFLNK:
+            return reparseTag == IO_REPARSE_TAG_SYMLINK;
+        case PY_IFMNT:
+            return reparseTag == IO_REPARSE_TAG_MOUNT_POINT;
+        case PY_IFLRP:
+            return IsReparseTagNameSurrogate(reparseTag);
+        case PY_IFRRP:
+            return reparseTag && !IsReparseTagNameSurrogate(reparseTag);
```

Comment:
It looks like we have this mapping 3 times. Should we use a helper function?

Reply:
An inline function could be defined that takes `testedType`, `attributes`, and `reparseTag`.